File: 0724-find-pivot-index/solution.java
class Solution {
    public int pivotIndex(int[] nums) {
        int n = nums.length;
        int totalSum = 0;

        for(int i=0; i<n; i++) {
            totalSum += nums[i];
        }
        /*
        👉 Calculate total sum once.
            Then as you iterate:
                •	Keep track of leftSum
                •	Derive rightSum from total

                totalSum = sum of all elements
                At index i, the array is split into three parts:
                [ left part ]  nums[i]  [ right part ]
                totalSum = leftSum + nums[i] + rightSum

                //REARRANGE to:
                We want rightSum.
                rightSum = totalSum - leftSum - nums[i]
        */
        int leftSum = 0;
        int rightSum = 0;

        for(int i=0; i<n; i++) {
            rightSum = totalSum - leftSum - nums[i];

            if(leftSum == rightSum) {
                return i;
            }
            leftSum += nums[i];
        }
        return -1;
    }
}

File: 0724-find-pivot-index/solution.swift
class Solution {
    func pivotIndex(_ nums: [Int]) -> Int {
        var n = nums.count
        var leftSum = Array(repeating: 0, count: n)
        var rightSum = Array(repeating: 0, count: n)

        leftSum[0] = nums[0]
        for i in 1..<nums.count {
            leftSum[i] = leftSum[i-1] + nums[i]
        }
        rightSum[n-1] = nums[n-1]
        for i in stride(from: n-2, through: 0, by: -1) {
            rightSum[i] = rightSum[i+1] + nums[i]
        }

        for i in 0..<nums.count {
            // print(leftSum[i] , rightSum[i])
            if leftSum[i] - nums[i] == rightSum[i] - nums[i] {
                return i
            }
        }

        return -1
    }
}
